Improvments to REPL tab-completion (#3227)

rndnoise opened the pull request:
- Complete all names that have been imported (transitively or directly)
- Do not complete names that haven't been imported
- Only recompute list of names after import or adding a let binding
  rather than after each request for name completion

This pull request fixes #3227

hdgarrood commented on src/Language/PureScript/Interactive/Types.hs:
```diff
@@ -54,17 +64,47 @@ allImportsOf m PSCiState{psciImportedModules = is} =
 
 -- * State helpers
 
+updateImportExports :: PSCiState -> PSCiState
```
Could you add a comment here saying something like

> This function updates the Imports and Exports values in the PSCiState, which are used for handling completions. This function must be called whenever the PSCiState is modified to ensure that completions remain accurate.

hdgarrood replied:
Alternatively, is it worth using an explicit export list and hiding the `PSCiState` constructor so that we can easily verify that the `PSCiState` can't get out of sync?

rndnoise replied:
Yeah, I think both are good ideas.

hdgarrood commented on src/Language/PureScript/Interactive/Types.hs:
```diff
@@ -25,10 +30,15 @@ data PSCiState = PSCiState
 
 initialPSCiState :: PSCiState
-initialPSCiState = PSCiState [] [] []
+initialPSCiState = updateImportExports (PSCiState [] [] [] nullImports nullExports)
```
Will `updateImportExports` do anything here?

rndnoise replied:
Yes, perhaps surprisingly. It adds the types from `Prim`, which is implicitly imported in the module created in that function.

hdgarrood commented on tests/TestPsci/CompletionTest.hs:
```diff
@@ -73,16 +78,13 @@ completionTestData supportModuleNames =
+  , ("G", ["GT"])
+  , ("P.G", ["P.GT"])
+  , ("P.uni", ["P.unit"])
+  , ("voi", []) -- import Prelude exclude (void)
```
The syntax is `import Prelude hiding (void)`

rndnoise commented:
I've hidden the `PSCiState` constructor and added a comment to `updateImportExports`.

I also made an optimization to `tests/TestPsci/CompletionTest.hs` so the module is reused rather than setup and recompiled for each assertion (running the REPL tests was noticeably slow, it's much faster now). The remaining slow bit is `tests/TestPsci/CommandTest.hs` which I've left alone.

rndnoise commented:
The build is failing due to an unused function `allImportsOf`. It's no longer used as the result of this PR, but I didn't want to delete someone's work without their permission. I think this code is yours, @hdgarrood ... should I remove it?

rndnoise commented:
After a closer look, I don't think `allImportsOf` was useful outside of the older tab-completion code that called it; so I went ahead and removed it so the build would pass. Let me know if any further changes are needed before merging.

hdgarrood commented on src/Language/PureScript/Interactive.hs:
```diff
@@ -190,9 +190,8 @@ handleShowImportedModules
-  print' $ showModules importedModules
-  return ()
+  st <- get
+  print' $ showModules (psciImportedModules st)
```
Bit of a nitpick, but using `$` as well as parens on the same line is stylistically perhaps not the best thing to do. I don't mind which but would you mind sticking to one or the other?

rndnoise commented:
OK, I think I've addressed all of your points.

> So would you mind splitting that off for now?

Sure, it is done. In general though, how do contributors here manage related or interdependent pull requests? Should those not be made into PRs until the dependency has been merged, or do people create PRs that have stacked on other PR branches?

> I just realised it would probably be helpful to add a comment near initialPSCiState to clarify the reason we need to use updateImportsExports there?

I've removed that call and exported `Sugar.Names.Env.primExports` so that it could be used here instead of `nullExports`.

> Also, looking at this I realised that we haven't actually managed to guarantee that the Imports and Exports stay in sync this way;

Good catch. I saw a compiler error halfway through making that change and assumed it confirmed my guess that I could export the type without exporting the constructor, but it must have been a different error. It is now a data declaration with some extra "getter" methods, so now it shouldn't be possible to take apart or build a `PSCiState` directly.

> Bit of a nitpick, but using $ as well as parens on the same line is stylistically perhaps not the best thing to do

No problem.

kRITZCREEK commented on src/Language/PureScript/Interactive/Types.hs:
```diff
@@ -22,13 +49,33 @@ data PSCiConfig = PSCiConfig
-  , psciLoadedExterns       :: [(P.Module, P.ExternsFile)]
-  } deriving Show
+  [ImportedModule]            -- ^ psciImportedModules
+  [P.Declaration]             -- ^ psciLetBindings
```
Haddock is crashing on this line. It doesn't seem like you're allowed to annotate the individual arguments to a constructor, just the whole constructor or record fields: https://www.haskell.org/haddock/doc/html/ch03s02.html

rndnoise replied:
Ah, thanks

hdgarrood commented on src/Language/PureScript/Interactive/Types.hs:
```diff
@@ -1,11 +1,38 @@
+module Language.PureScript.Interactive.Types
+  ( PSCiConfig(..)
+  , PSCiState -- constructor is not exported, to prevent psciImports and psciExports from
+              -- becoming inconsistent with importedModules, letBindnigs and loadedExterns
```
typo here? "letBindnigs"

rndnoise replied:
Yes, thanks :-)

hdgarrood commented:
This looks great; I'm happy to merge once CI is passing.
